allennlp/data/dataset_readers/quora_paraphrase.py

```diff
@@ -0,0 +1,85 @@
+from typing import Dict
+import re
+import logging
+import csv
+import zipfile
+import io
+
+from overrides import overrides
+
+from allennlp.common.file_utils import cached_path
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import LabelField, TextField, Field
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Tokenizer, WordTokenizer
+from allennlp.data.tokenizers.word_splitter import JustSpacesWordSplitter
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def parse_file_uri(uri: str):
+    match = re.fullmatch(r'\((.*)\)#(.*)', uri)      # pylint: disable=anomalous-backslash-in-string
+    if match and len(match.groups()) == 2:
+        return match.groups()[0], match.groups()[1]
+    else:
+        return uri, None
+
+@DatasetReader.register("quora_paraphrase")
+class QuoraParaphraseDatasetReader(DatasetReader):
+    """
+    Reads a Quora paraphrase data
+
+    Parameters
+    ----------
+    lazy : ``bool`` (optional, default=False)
+        Passed to ``DatasetReader``.  If this is ``True``, training will start sooner, but will
+        take longer per batch.  This also allows training with datasets that are too large to fit
+        in memory.
+    tokenizer : ``Tokenizer``, optional
+        Tokenizer to use to split the premise and hypothesis into words or other kinds of tokens.
+        Defaults to ``WordTokenizer(JustSpacesWordSplitter())``.
+    token_indexers : ``Dict[str, TokenIndexer]``, optional
+        Indexers used to define input token representations. Defaults to ``{"tokens":
+        SingleIdTokenIndexer()}``.
+    """
+    def __init__(self,
+                 lazy: bool = False,
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        super().__init__(lazy)
+        self._tokenizer = tokenizer or WordTokenizer(JustSpacesWordSplitter())
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+    @overrides
+    def _read(self, file_path):
+        logger.info("Reading instances from lines in file at: %s", file_path)
+        file_name, member = parse_file_uri(file_path)
```

Comment:
You've duplicated the logic below between the two cases (where `member is None` and where it isn't).  Seems like it would be cleaner to push the file decision logic into this method.  So you'd have something like:

```python
def open_uri(file_path):
    file_name, member = parse_file_uri(file_path)
    file_path = cached_path(file_path)
    if member is None:
        return open(file_path, 'r')
    else:
        # zipfile opening logic here

def _read(self, file_path):
    with open_uri(file_path) as data_file:
       tsvin = ...
```

@joelgrus probably knows more than I do about how to handle the `with` stuff properly in this case.

Reply:
@matt-gardner @joelgrus Sure I will try to refactor this. One related question is I am currently using my personal s3 to host the data file `https://s3-us-west-1.amazonaws.com/handsomezebra/public/Quora_question_pair_partition.zip`. Do you think you can help me upload the files to allennlp's s3?

Reply:
Yes, I'd be happy to put that data file (and any model files you have) into our s3 bucket. Just tell me what all of them are.

Reply:
Just this data file for now https://s3-us-west-1.amazonaws.com/handsomezebra/public/Quora_question_pair_partition.zip
Thanks!

Reply:
Ok, I unzipped it and put the train, dev and test files here: `https://s3-us-west-1.amazonaws.com/allennlp/datasets/quora-question-paraphrase/{train,dev,test}.tsv`, along with the README.  I left out the word vector file.

Reply:
(Unzipping it also makes it so you can just remove the zip file logic and simplify that piece of the code.  I'd vote for handling zip files in a more general way in our `common` module, which we can certainly do, but is probably more appropriate for a separate PR.)

Reply:
Sure I will remove the zip file reading. 
I tried the s3 URL https://s3-us-west-1.amazonaws.com/allennlp/datasets/quora-question-paraphrase/train.tsv but got a redirection error. But this URL is fine: https://allennlp.s3-us-west-2.amazonaws.com/datasets/quora-question-paraphrase/train.tsv. I can use the second URL but other data on allennlp doesn't seem to have this problem.

Reply:
Ah, sorry, that's my fault, I had a copy-paste error in the link I gave you.